plugins/inputs/docker_log/docker_logs.go

```diff
@@ -0,0 +1,440 @@
+  docker_label_include = []
+  docker_label_exclude = []
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+`
+
+func (d *DockerLogs) Description() string {
+	return "Plugin to get docker logs"
+}
+
+func (d *DockerLogs) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *DockerLogs) Gather(acc telegraf.Accumulator) error {
+	/*Check to see if any new containers have been created since last time*/
+	d.containerListUpdate(acc)
+	return nil
+}
+
+/*Following few functions have been inherited from telegraf docker input plugin*/
+func (d *DockerLogs) createContainerFilters() error {
+	filter, err := filter.NewIncludeExcludeFilter(d.ContainerInclude, d.ContainerExclude)
+	if err != nil {
+		return err
+	}
+	d.containerFilter = filter
+	return nil
+}
+
+func (d *DockerLogs) createLabelFilters() error {
+	filter, err := filter.NewIncludeExcludeFilter(d.LabelInclude, d.LabelExclude)
+	if err != nil {
+		return err
+	}
+	d.labelFilter = filter
+	return nil
+}
+
+func (d *DockerLogs) createContainerStateFilters() error {
+	if len(d.ContainerStateInclude) == 0 && len(d.ContainerStateExclude) == 0 {
+		d.ContainerStateInclude = []string{"running"}
+	}
+	filter, err := filter.NewIncludeExcludeFilter(d.ContainerStateInclude, d.ContainerStateExclude)
+	if err != nil {
+		return err
+	}
+	d.stateFilter = filter
+	return nil
+}
+
+func (d *DockerLogs) addToContainerList(containerId string, logReader io.ReadCloser) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.containerList[containerId] = logReader
+	return nil
+}
+
+func (d *DockerLogs) removeFromContainerList(containerId string) error {
+	delete(d.containerList, containerId)
+	return nil
+}
+
+func (d *DockerLogs) containerInContainerList(containerId string) bool {
+	if _, ok := d.containerList[containerId]; ok {
+		return true
+	}
+	return false
+}
+
+func (d *DockerLogs) stopAllReaders() error {
+	for _, container := range d.containerList {
+		container.Close()
+	}
+	return nil
+}
+
+func (d *DockerLogs) containerListUpdate(acc telegraf.Accumulator) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout.Duration)
+	defer cancel()
+	if d.client == nil {
+		log.Printf("%s : Dock client is null", ERR_PREFIX)
+		return nil
+	}
+	containers, err := d.client.ContainerList(ctx, d.opts)
+	if err != nil {
+		log.Printf("%s : %s ", ERR_PREFIX, err.Error())
+		return err
+	}
+	for _, container := range containers {
+		if d.containerInContainerList(container.ID) {
+			continue
+		}
+		d.wg.Add(1)
+		/*Start a new goroutine for every new container that has logs to collect*/
+		go func(c types.Container) {
+			defer d.wg.Done()
+			err := d.getContainerLogs(c, acc)
+			if err != nil {
+				log.Printf("%s : %s ", ERR_PREFIX, err.Error())
+			}
+			d.removeFromContainerList(c.ID)
```

Comment:
Must lock d.containerList because this is called inside a goroutine that is not waited for.

Reply:
Added d.mu.Lock and unlock inside the removeFromContainerList . Is there anything I should be adding?

Reply:
I think this is sufficient, the code that can run concurrently is the container goroutines and the `Stop()` function.